### PR TITLE
Create _get_pools method for Antminer backend class

### PR DESCRIPTION
**Pull request by Ytemiloluwa**
This method creates the _get_pools  function for the antminer type. @b-rowan

### CODE REVIEW

File: pyasic/miners/backends/antminer.py [hunk 1]
```diff
@@ -32,6 +31,8 @@
 from pyasic.rpc.antminer import AntminerRPCAPI
 from pyasic.ssh.antminer import AntminerModernSSH
 from pyasic.web.antminer import AntminerModernWebAPI, AntminerOldWebAPI
+from pyasic.data.pools import PoolMetrics
+from pyasic.errors import APIError
 
 ANTMINER_MODERN_DATA_LOC = DataLocations(
```
Comment (b-rowan) [inline pyasic/miners/backends/antminer.py:37]: Make sure to add the kwargs key to the data locations dict.

**Review by b-rowan — CHANGES_REQUESTED** [review]

**Review by b-rowan — COMMENTED** [review]
Pretty sure this is fine, but I don't really have much in the way of stock antminers.

Tentatively approved, but @s0kil do you mind just testing that this works using both `get_data` and `get_pools` on a real miner?

File: pyasic/miners/backends/antminer.py [hunk 2]
```diff
@@ -351,6 +352,35 @@ async def _get_uptime(self, rpc_stats: dict = None) -> Optional[int]:
...
+            try:
+                pools = rpc_pools.get("POOLS", [])
+                for index, pool_info in enumerate(pools):
+                    pool_data = PoolMetrics(
+                        accepted=pool_info.get("Accepted"),
+                        rejected=pool_info.get("Rejected"),
+                        get_failures=pool_info.get("Get Failures"),
+                        remote_failures=pool_info.get("Remote Failures"),
+                        active=pool_info.get("Stratum Active"),
+                        alive=pool_info.get("Alive"),
```
Comment (b-rowan) [inline pyasic/miners/backends/antminer.py:377]: Alive appears to not be a boolean value like Stratum Active, but shown in Status as "Alive":
```
alive=pool_info.get("Status") == "Alive",
```

File: pyasic/miners/backends/antminer.py [hunk 3]
```diff
@@ -351,6 +352,35 @@ async def _get_uptime(self, rpc_stats: dict = None) -> Optional[int]:
...
+                        url=pool_info.get("URL"),
+                        user=pool_info.get("User"),
+                        index=pool_info.get("Index")
```
Comment (b-rowan) [inline pyasic/miners/backends/antminer.py:380]: Use the "POOL" key, not the "Index".  Not sure why they keyed it this way, but that's the right value.
```
index=pool_info.get("POOL"),
```

**Review by b-rowan — CHANGES_REQUESTED** [review]

**Comment by s0kil** [comment]
```bash
$ curl --digest -u root:root http://192.12.1.188/cgi-bin/pools.cgi | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   351  100   351    0     0   3714      0 --:--:-- --:--:-- --:--:--  3734
100  1019    0  1019    0     0   4468      0 --:--:-- --:--:-- --:--:--  4468
{
  "STATUS": {
    "STATUS": "S",
    "when": 1716959474,
    "Msg": "pools",
    "api_version": "1.0.0"
  },
  "INFO": {
    "miner_version": "49.0.1.3",
    "CompileTime": "Tue Dec  6 16:12:30 CST 2022",
    "type": "Antminer S19 Pro"
  },
  "POOLS": [
    {
      "index": 0,
      "url": "stratum+tcp://fake.com:3334",
      "user": "fake.user1",
      "status": "Dead",
      "priority": 0,
      "getworks": 0,
      "accepted": 0,
      "rejected": 0,
      "discarded": 0,
      "stale": 0,
      "diff": "",
      "diff1": 0,
      "diffa": 0,
      "diffr": 0,
      "diffs": 0,
      "lsdiff": 0,
      "lstime": "0"
    },
    {
      "index": 1,
      "url": "stratum+tcp://fake.com:3334",
      "user": "fake.user1",
      "status": "Dead",
      "priority": 1,
      "getworks": 0,
      "accepted": 0,
      "rejected": 0,
      "discarded": 0,
      "stale": 0,
      "diff": "",
      "diff1": 0,
      "diffa": 0,
      "diffr": 0,
      "diffs": 0,
      "lsdiff": 0,
      "lstime": "0"
    },
    {
      "index": 2,
      "url": "stratum+tcp://fake.com:3334",
      "user": "fake.user1",
      "status": "Dead",
      "priority": 2,
      "getworks": 0,
      "accepted": 0,
      "rejected": 0,
      "discarded": 0,
      "stale": 0,
      "diff": "",
      "diff1": 0,
      "diffa": 0,
      "diffr": 0,
      "diffs": 0,
      "lsdiff": 0,
      "lstime": "0"
    }
  ]
}
```

```go
type Pools struct {
	Info   Info           `json:"INFO"`
	Pools  []Pool         `json:"POOLS"`
	Status ResponseStatus `json:"STATUS"`
}

type Info struct {
	MinerVersion string `json:"miner_version"`
	CompileTime  string `json:"CompileTime"`
	Type         string `json:"type"`
}

type Pool struct {
	Diff      string `json:"diff"`
	URL       string `json:"url"`
	User      string `json:"user"`
	Status    string `json:"status"`
	LsTime    string `json:"lstime"`
	Discarded int64  `json:"discarded"`
	Accepted  int64  `json:"accepted"`
	Rejected  int64  `json:"rejected"`
	Index     int64  `json:"index"`
	Stale     int64  `json:"stale"`
	GetWorks  int64  `json:"getworks"`
	Diff1     int64  `json:"diff1"`
	DiffA     int64  `json:"diffa"`
	DiffR     int64  `json:"diffr"`
	Diffs     int64  `json:"diffs"`
	LsDiff    int64  `json:"lsdiff"`
	Priority  int64  `json:"priority"`
}
```

**Comment by b-rowan** [comment]
This seems different than than the RPC endpoint?  Like my simulator has way more information there...  My guess is that the pools endpoint is using the `new_api` flag?

**Comment by s0kil** [comment]
Well, I've also tried the 4028 API, the response seems similar:

```bash
nc 192.12.1.188 4028
{"command": "pools"}
```


```json
{
  "STATUS": [
    {
      "STATUS": "S",
      "When": 1717000009,
      "Code": 7,
      "Msg": "3 Pool(s)",
      "Description": "cgminer 1.0.0"
    }
  ],
  "POOLS": [
    {
      "POOL": 0,
      "URL": "stratum+tcp://test.xyz:3334",
      "Status": "Alive",
      "Priority": 0,
      "Quota": 1,
      "Long Poll": "N",
      "Getworks": 1102,
      "Accepted": 7533,
      "Rejected": 5,
      "Discarded": 3594174,
      "Stale": 0,
      "Get Failures": 0,
      "Remote Failures": 0,
      "User": "test.user1",
      "Last Share Time": "0:00:03",
      "Diff": "65.5K",
      "Diff1 Shares": 0,
      "Proxy Type": "",
      "Proxy": "",
      "Difficulty Accepted": 1020543787.5,
      "Difficulty Rejected": 655350.0,
      "Difficulty Stale": 0.0,
      "Last Share Difficulty": 65535.0,
      "Has Stratum": true,
      "Stratum Active": true,
      "Stratum URL": "test.xyz",
      "Has GBT": false,
      "Best Share": 5468855623.0,
      "Pool Rejected%": 0.0,
      "Pool Stale%%": 0.0
    },
    {
      "POOL": 1,
      "URL": "stratum+tcp://test.xyz:3334",
      "Status": "Alive",
      "Priority": 1,
      "Quota": 1,
      "Long Poll": "N",
      "Getworks": 1,
      "Accepted": 0,
      "Rejected": 0,
      "Discarded": 0,
      "Stale": 0,
      "Get Failures": 0,
      "Remote Failures": 0,
      "User": "test.user1",
      "Last Share Time": "0",
      "Diff": "",
      "Diff1 Shares": 0,
      "Proxy Type": "",
      "Proxy": "",
      "Difficulty Accepted": 0.0,
      "Difficulty Rejected": 0.0,
      "Difficulty Stale": 0.0,
      "Last Share Difficulty": 0.0,
      "Has Stratum": true,
      "Stratum Active": false,
      "Stratum URL": "",
      "Has GBT": false,
      "Best Share": 0.0,
      "Pool Rejected%": 0.0,
      "Pool Stale%%": 0.0
    },
    {
      "POOL": 2,
      "URL": "",
      "Status": "Deed",
      "Priority": 2,
      "Quota": 1,
      "Long Poll": "N",
      "Getworks": 0,
      "Accepted": 0,
      "Rejected": 0,
      "Discarded": 0,
      "Stale": 0,
      "Get Failures": 0,
      "Remote Failures": 0,
      "User": "",
      "Last Share Time": "0",
      "Diff": "",
      "Diff1 Shares": 0,
      "Proxy Type": "",
      "Proxy": "",
      "Difficulty Accepted": 0.0,
      "Difficulty Rejected": 0.0,
      "Difficulty Stale": 0.0,
      "Last Share Difficulty": 0.0,
      "Has Stratum": false,
      "Stratum Active": false,
      "Stratum URL": "",
      "Has GBT": false,
      "Best Share": 0.0,
      "Pool Rejected%": 0.0,
      "Pool Stale%%": 0.0
    }
  ],
  "id": 1
}
```

**Comment by s0kil** [comment]
> This seems different than than the RPC endpoint? Like my simulator has way more information there... My guess is that the pools endpoint is using the `new_api` flag?

It may be version differences, 2021 FW and 2022 do have multiple major changes.

File: pyasic/miners/backends/antminer.py (hunk 2 above)
Comment (Ytemiloluwa) [reply pyasic/miners/backends/antminer.py:377]: Okay, I will push an update soon.

File: pyasic/miners/backends/antminer.py (hunk 3 above)
Comment (Ytemiloluwa) [reply pyasic/miners/backends/antminer.py:380]: okay, Noted. ✅

**Review by b-rowan — CHANGES_REQUESTED** [review]
Everything here looks good, but its not showing up in `MinerData`.  You wioll just need to add `pools: list[PoolMetrics] = field(default_factory=list)` into the `MinerData` dataclass, located at `pyasic/data/__init__.py`

**Comment by b-rowan** [comment]
> Everything here looks good, but its not showing up in `MinerData`. You wioll just need to add `pools: list[PoolMetrics] = field(default_factory=list)` into the `MinerData` dataclass, located at `pyasic/data/__init__.py`

And just make sure to put it into the docstring as well.

**Comment by Ytemiloluwa** [comment]
> > Everything here looks good, but its not showing up in `MinerData`. You wioll just need to add `pools: list[PoolMetrics] = field(default_factory=list)` into the `MinerData` dataclass, located at `pyasic/data/__init__.py`
> 
> And just make sure to put it into the docstring as well.

Got it ✅

**Pull request merged**